jwst/assign_wcs/nirspec.py

```diff
@@ -662,31 +637,66 @@ def get_open_msa_slits(msa_file, msa_metadata_id, dither_position,
+            source_id = slitlet_id
+
+            # Hardwire the source info for background slits, because there's
+            # no source info for them in the msa_file
             source_xpos = 0.5
             source_ypos = 0.5
-            source_id = _get_bkg_source_id(bkg_counter, max_source_id)
-            log.info(f'Slitlet_id {slitlet_id} is background only; assigned source_id = {source_id}')
-            bkg_counter += 1
+            source_name = "background_{}".format(slitlet_id)
+            source_alias = "bkg_{}".format(slitlet_id)
+            stellarity = 0.0
+            source_ra = 0.0
+            source_dec = 0.0
+            log.info(f'Slitlet {slitlet_id} is background only; assigned source_id={source_id}')
 
-        # There is 1 main shutter: phew, that makes it easier.
+        # There is 1 main shutter: this is a slit containing either a real or virtual source
         elif n_main_shutter == 1:
             xcen, ycen, quadrant, source_xpos, source_ypos = [
                 (s['shutter_row'], s['shutter_column'], s['shutter_quadrant'],
                  s['estimated_source_in_shutter_x'],
                  s['estimated_source_in_shutter_y'])
                 for s in slitlets_sid if s['background'] == 'N'][0]
+            shutter_id = xcen + (ycen - 1) * 365  # shutter numbers in MSA file are 1-indexed
 
             # y-size
             jmin = min([s['shutter_column'] for s in slitlets_sid])
             jmax = max([s['shutter_column'] for s in slitlets_sid])
             j = ycen
             ymax = yhigh + margin + (jmax - j) * 1.15
             ymin = -(-ylow + margin) + (jmin - j) * 1.15
+
             # get the source_id from the primary shutter entry
             for i in range(len(slitlets_sid)):
                 if slitlets_sid[i]['primary_source'] == 'Y':
                     source_id = slitlets_sid[i]['source_id']
 
+            # Slits with a real source assigned have a source_id > 0
+            if source_id > 0:
+                # Get source info for this normal slitlet
+                try:
+                    source_name, source_alias, stellarity, source_ra, source_dec = [
+                        (s['source_name'], s['alias'], s['stellarity'], s['ra'], s['dec'])
+                        for s in msa_source if s['source_id'] == source_id][0]
+                except IndexError:
+                    log.warning("Could not retrieve source info from MSA file")
+
+            # Slits with source_id < 0 are "virtual" slits, with no source assigned
+            else:
+                # Hardwire the source info for this virtual slit, because there's none in the MSA file
+                source_xpos = 0.5
+                source_ypos = 0.5
+                source_name = "virtual_{}".format(abs(source_id))
+                source_alias = "vrt_{}".format(abs(source_id))
+                stellarity = 0.0
+                source_ra = 0.0
+                source_dec = 0.0
```

Comment:
Wow, you're right!  I was (blindly) assuming that slits with a virtual source, identified by their negative source_id value in the MSA shutter table, would not have any corresponding info in the MSA source table, and hence was not even bothering to load it. But they *do* have entries in the source table. For example, here are the shutter table rows for virtual slit 87, which has source_id=-70. And there is a corresponding row in the source table for source_id=-70:

```
FITS_rec([(87, 88, 4, 345, 1, -70, 'N', 'OPEN', 0.5, 0.5, 1, 'Y'),
          (87, 88, 4, 345, 1,   0, 'Y', 'OPEN', nan, nan, 2, 'N'),
          (87, 88, 4, 345, 1,   0, 'Y', 'OPEN', nan, nan, 3, 'N')],

(1345, -70, '1345_-70', '-09:-039:0.0000 +53:00:57.29', 215.1912160802715, 53.01591471620116, 'None', 0.0)
```
The x/ypos values are set to 0.5 in the shutter table, so that won't have any change relative to the defaults I had been assigning, but it would result in real values for all the other items, like source_name (1345_-70), alias, ra, dec, and stellarity. 

Good catch!!

Reply:
Latest commits have updated this function to get source info from the MSA source table for both "real" (source_id > 0) and virtual (source_id < 0) sources. No more zeroing out source-related info for slits with virtual sources.